File: sparqlGraphWeb/sparqlGraph/js/sparqlconnection.js
/**
 ** Copyright 2016-17 General Electric Company
 **
 ** Authors:  Paul Cuddihy, Justin McHugh
 **
 ** Licensed under the Apache License, Version 2.0 (the "License");
 ** you may not use this file except in compliance with the License.
 ** You may obtain a copy of the License at
 **
 **     http://www.apache.org/licenses/LICENSE-2.0
 **
 ** Unless required by applicable law or agreed to in writing, software
 ** distributed under the License is distributed on an "AS IS" BASIS,
 ** WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 ** See the License for the specific language governing permissions and
 ** limitations under the License.
 */

/*
 * Connection allowing Ontology to be in different place(s) than the data
 * "Interface" for historical reasons is an endpoint plus "graph" or "dataset"
 */

var SparqlConnection = function(jsonText) {

	//--used to support deprecated functions ---
	this.depServerType ="";
	this.depDomain="";
	this.enableOwlImports = false;
	//--old attributes still called by some legacy users ---
	this.serverType = "";

	this.dataServerUrl = "";
	this.dataKsServerURL = "";
	this.dataSourceDataset = "";

	this.ontologyServerUrl = "";
	this.ontologyKsServerURL = "";
	this.ontologySourceDataset = "";

	//-------------------------------------

	// The actual function:
	if (jsonText) {
    	this.fromString(jsonText);
    } else {
    	this.name = "";
    	this.domain = "";

    	this.modelInterfaces = [];
    	this.dataInterfaces = [];
    }
};


SparqlConnection.prototype = {

	toJson : function () {
        // backwards-compat fix.
        if ((typeof this.enableOwlImports == undefined)) {
            this.enableOwlImports = false;
        }

		var jObj = {
			name: this.name,
            // this is no longer editable in SPARQLgraph and defaults to ""
            // it is retained for backwards compatibility
			domain: this.domain,
            enableOwlImports: this.enableOwlImports,
			model: [],
			data: []
		};

		// add model interfaces
		for (var i=0; i < this.modelInterfaces.length; i++) {
			var mi = this.modelInterfaces[i];
			jObj.model.push({
					type: mi.getServerType(),
					url: mi.getServerURL(),
					graph: mi.getGraph()
			});
		}

		// add data interfaces
		for (var i=0; i < this.dataInterfaces.length; i++) {
			var di = this.dataInterfaces[i];
			jObj.data.push({
					type: di.getServerType(),
					url: di.getServerURL(),
					graph: di.getGraph()
			});
		}

		return jObj;
	},

	fromJson : function (jObj) {

		this.name = jObj.name;
		this.domain = jObj.domain;
        // be extra-safe as this is removed except for backwards-compatibility
        if (! this.domain) { this.domain = ""; }

		this.modelInterfaces = [];
    	this.modelDomains = [];
    	this.modelNames = [];

    	this.dataInterfaces = [];
    	this.dataNames = [];

		if (jObj.hasOwnProperty("dsURL")) {

			// backwards compatible read
			console.log("SparqlConnection.fromJson() is reading old-fashioned connection JSON.")

			// If any field doesn't exist, presume it exists in the other connection
			this.addModelInterface(jObj.type,
								  jObj.hasOwnProperty("onURL") ? jObj.onURL : jObj.dsURL,
								  jObj.hasOwnProperty("onDataset") ? jObj.onDataset : jObj.dsDataset
								  );
			this.addDataInterface(jObj.type,
								  jObj.hasOwnProperty("dsURL") ? jObj.dsURL : jObj.onURL,
								  jObj.hasOwnProperty("dsDataset") ? jObj.dsDataset : jObj.onDataset
							      );
		} else {
			// normal read

			// read model interfaces
	    	for (var i=0; i < jObj.model.length; i++) {
	    		var m = jObj.model[i];
                var graph = m.hasOwnProperty("dataset") ? m.dataset : m.graph;
	    		this.addModelInterface(m.type, m.url, graph);
	    	}
	    	// read data interfaces
	    	for (var i=0; i < jObj.data.length; i++) {
	    		var d = jObj.data[i];
                var graph = d.hasOwnProperty("dataset") ? d.dataset : d.graph;
	    		this.addDataInterface(d.type, d.url, graph);
	    	}
		}

        if (jObj.hasOwnProperty("enableOwlImports")) {
            this.enableOwlImports = jObj.enableOwlImports;
        } else {
            this.enableOwlImports = false;
        }
		this.fillDeprecatedFields();
	},

	fromString : function (jsonText) {
		jObj = JSON.parse(jsonText);
		this.fromJson(jObj);
	},

	toString : function () {
		return JSON.stringify(this.toJson());
	},

	equals : function (other, ignoreName) {
		var thisStr = this.toString();
		var otherStr = other.toString();
		if (ignoreName) {
			thisStr = thisStr.replace(this.name, "NAME");
			otherStr = otherStr.replace(other.name, "NAME");
		}
		return (thisStr == otherStr);
	},

	//== New-fashioned way to build a Connection ==
	setName : function (name) {
		this.name = name;
	},

	setDomain : function (domain) {
		this.domain = domain
	},

	addModelInterface : function (sType, url, graph) {
		this.modelInterfaces.push(this.createInterface(sType, url, graph));
	},

	addDataInterface : function (sType, url, graph) {
		this.dataInterfaces.push(this.createInterface(sType, url, graph));
	},

	delModelInterface : function (i) {
		this.modelInterfaces.splice(i, 1);
	},

	delDataInterface : function (i) {
		this.dataInterfaces.splice(i, 1);
	},

    isOwlImportsEnabled : function() {
        return this.enableOwlImports;
    },

    setOwlImportsEnabled : function(enableOwlImports) {
		this.enableOwlImports = enableOwlImports;
	},
	getModelInterfaceCount : function () {
		return this.modelInterfaces.length;
	},
	getModelInterface : function (i) {
		return this.modelInterfaces[i];
	},
	getDomain : function () {
		return this.domain;
	},
	getName : function() {
		return this.name;
	},
	getDataInterfaceCount : function () {
		return this.dataInterfaces.length;
	},
	getDataInterface : function (i) {
		if (typeof i == "undefined") {
			// DEPRECATED: old and new name are the same, but old had no param
			console.log("NOTE: called deprecated SparqlConnection.getDataInterface() with no param");
			return this.dataInterfaces[0];
		} else {
			return this.dataInterfaces[i];
		}

	},

	getDefaultQueryInterface : function () {
		if (this.dataInterfaces.length > 0) {
			return this.dataInterfaces[0];
		} else if (this.modelInterfaces.length > 0) {
			return this.modelInterfaces[0];
		} else {
			throw new Error("This SparqlConnection has no endpoints.");
		}
	},

	getInsertInterface : function () {
		if (this.dataInterfaces.length == 1) {
			return this.dataInterfaces.get[0];
		} else {
			throw new Error("Expecting one data endpoint for INSERT.  Found " + this.dataInterfaces.length);
		}
	},

	// Is number of endpoint serverURLs == 1
	isSingleServerURL : function() {
		var url = "";
		for (var i=0; i < this.modelInterfaces.length; i++) {
			var e =  this.modelInterfaces[i].getServerURL();
			if (url == "") {
				url = e;
			} else if (e != url) {
				return false;
			}
		}

		// add data interfaces
		for (var i=0; i < this.dataInterfaces.length; i++) {
			var e =  this.dataInterfaces[i].getServerURL();
			if (url == "") {
				url = e;
			} else if (e != url) {
				return false;
			}
		}

		// if there are no serverURLs then false
		if (url == "") {
			return false;
		} else {
			return true;
		}
	},

	// get list of graphs for a given serverURL
	getGraphsForServer : function(serverURL) {
		var ret = [];

		for (var i=0; i < this.modelInterfaces.length; i++) {
			var e =  this.modelInterfaces[i];
			if (e.getServerURL() == serverURL &&  ret.indexOf(e.getGraph()) == -1) {
				ret.push(e.getGraph());
			}
		}

		for (var i=0; i < this.dataInterfaces.length; i++) {
			var e =  this.dataInterfaces[i];
			if (e.getServerURL() == serverURL &&  ret.indexOf(e.getGraph()) == -1) {
				ret.push(e.getGraph());
			}
		}

		return ret;
	},

	//---------- private function
	createInterface : function (stype, url, graph) {
        return new SparqlServerInterface(stype, url, graph);
	},

	// DEPRECATED private
	fillDeprecatedFields : function () {
		// make deprecated things "work" using first of each interface
		if (this.modelInterfaces.length < 1 || this.dataInterfaces.length < 1) {
			return;
		}

		this.serverType = this.modelInterfaces[0].getServerType()
		this.ontologyServerUrl = this.modelInterfaces[0].getServerURL();
		this.ontologySourceDataset = this.modelInterfaces[0].getDataset();
		this.dataServerUrl = this.dataInterfaces[0].getServerURL();
		this.dataSourceDataset = this.dataInterfaces[0].getDataset();

	},

	// DEPRECATED: kept for some legacy callers.
	build : function () {
		console.log("ERROR: Using DEPRECATED SparqlConnection functions.");
		console.log("	    Change to setName() addDataInterface() addModelInterface().");

		// creating a single model and single data endpoint presuming that the
		// caller has already set the old-fashioned attributes.
		this.setName(this.name);
		this.addModelInterface(	this.serverType,
								this.ontologyServerUrl,
								this.ontologySourceDataset,
								this.domain,
								""
							);
		this.addDataInterface(	this.serverType,
								this.dataServerUrl,
								this.dataSourceDataset,
								""
							);
	},

	// DEPRECATED
	setup(name, serverType, domain) {
		console.log("NOTE: called deprecated SparqlConnection.setup();");

		this.name = name;
		this.depServerType = serverType;
		this.depDomain = domain;
	},
	// DEPRECATED
	getOntologyInterface : function () {
		console.log("NOTE: called deprecated SparqlConnection.getOntologyInterface()");
		return this.modelInterfaces[0];
	},
	// DEPRECATED
	setOntologyInterface(url, dataset, unused_KsURL) {
		console.log("NOTE: called deprecated SparqlConnection.setOntologyInterface();");

		this.modelInterfaces = [this.createInterface(this.depServerType, url, dataset)];
		this.modelDomains= [this.depDomain];
		this.modelNames=[""];
	},
	// DEPRECATED
	setDataInterface(url, dataset, unused_KsURL) {
		console.log("Note: called deprecated SparqlConnection.setDataInterface();");

		this.dataInterfaces = [this.createInterface(this.depServerType, url, dataset)];
		this.dataNames=[""];
	}
};

File: connectionUtils/src/main/java/com/ge/research/semtk/sparqlX/SparqlConnection.java
/**
 ** Copyright 2016-2018 General Electric Company
 **
 **
 ** Licensed under the Apache License, Version 2.0 (the "License");
 ** you may not use this file except in compliance with the License.
 ** You may obtain a copy of the License at
 ** 
 **     http://www.apache.org/licenses/LICENSE-2.0
 ** 
 ** Unless required by applicable law or agreed to in writing, software
 ** distributed under the License is distributed on an "AS IS" BASIS,
 ** WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 ** See the License for the specific language governing permissions and
 ** limitations under the License.
 */


package com.ge.research.semtk.sparqlX;

import java.util.ArrayList;
import java.util.Arrays;

import org.json.simple.JSONArray;
import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;

public class SparqlConnection {
/**
 * SparqlConnection is a "SparqlGraph" connection consisting of multiple Sparql Endpoints
 *    1) Ontology Model
 *    2) Data
 * 
 * 
 */
	
	private final String KEY_ENABLE_OWL_IMPORTS = "enableOwlImports";
	
	private String name = null;
	private String domain = "";  // deprecated
	private ArrayList<SparqlEndpointInterface> modelInterfaces = null;
	private ArrayList<SparqlEndpointInterface> dataInterfaces = null;
	private boolean enableOwlImports = false;
	
	/**
	 * Empty constructor
	 */
	public SparqlConnection () {
		this.name = "";
		this.domain = "";
		this.modelInterfaces = new ArrayList<SparqlEndpointInterface>();
		this.dataInterfaces = new ArrayList<SparqlEndpointInterface>();
		this.enableOwlImports = false;
	}
	
	/**
	 * Construct from json text
	 * @param jsonText
	 * @throws Exception
	 */
	public SparqlConnection(String jsonText) throws Exception {
		this();
	    this.fromString(jsonText);
	}
	
	/**
	 * Constructor
	 * @param name
	 * @param serverType virtuoso|fuseki|neptune
	 * @param serverURL  http://localhost:3030/DATASET where some triplestores don't accept /DATASET
	 * @param dataset  - will be used for model and data
	 * @throws Exception
	 */
	public SparqlConnection(String name, String serverType, String serverURL, String dataset) throws Exception{
		this();
		this.name = name;
		this.addDataInterface(serverType, 
				serverURL,
				dataset);
		this.addModelInterface(serverType, 
				serverURL,
				dataset);
	}
	
	/**
	 * Constructor
	 * @param name
	 * @param sei for model and data
	 * @throws Exception
	 */
	public SparqlConnection(String name, SparqlEndpointInterface sei) throws Exception {
		this();
		this.name = name;
		this.addDataInterface(sei);
		this.addModelInterface(sei);
	}
	
	/**
	 * Constructor
	 * @param name
	 * @param modelSei - model endpoint
	 * @param dataSei - data endpoint
	 * @throws Exception
	 */
	public SparqlConnection(String name, SparqlEndpointInterface modelSei, SparqlEndpointInterface dataSei) throws Exception {
		this();
		this.name = name;
		this.addDataInterface(dataSei);
		this.addModelInterface(modelSei);
	}
	
	@Deprecated
	public SparqlConnection(String name, String serverType, String dataServicetURL, String knowledgeServiceURL, String dataset) throws Exception{
		this();
		this.name = name;
		this.addDataInterface(serverType, 
				dataServicetURL,
				dataset);
		this.addModelInterface(serverType, 
				dataServicetURL,
				dataset);
	}
	
	/**
	 * Old constructor including "domain"
	 * @param name
	 * @param serverType
	 * @param dataServicetURL
	 * @param knowledgeServiceURL
	 * @param dataset
	 * @param domain
	 * @throws Exception
	 */
	@Deprecated
	public SparqlConnection(String name, String serverType, String dataServicetURL, String knowledgeServiceURL, String dataset, String domain) throws Exception{
		this();
		this.name = name;
		this.domain = (domain == null) ? "" : domain;
		this.addDataInterface(serverType, 
				dataServicetURL,
				dataset);
		this.addModelInterface(serverType, 
				dataServicetURL,
				dataset);
	}
	
	/**
	 * build json
	 * @return
	 */
	@SuppressWarnings("unchecked")
	public JSONObject toJson() {
		JSONObject jObj = new JSONObject();
		jObj.put("name", name);
		jObj.put("domain", (domain == null) ? "" : domain);
		
		JSONArray model = new JSONArray();
		JSONArray data = new JSONArray();

		for (int i=0; i < this.modelInterfaces.size(); i++) {			
			model.add(this.modelInterfaces.get(i).toJson());
		}
		
		for (int i=0; i < this.dataInterfaces.size(); i++) {
			data.add(this.dataInterfaces.get(i).toJson());
		}
		
		jObj.put("model", model);
		jObj.put("data", data);
		
		jObj.put(KEY_ENABLE_OWL_IMPORTS, this.enableOwlImports);
		return jObj;
	}
	
	/**
	 * Instantiate empty object from json
	 * @param jObj
	 * @throws Exception
	 */
	public void fromJson(JSONObject jObj) throws Exception {
		
		if(jObj.entrySet().size() == 1 && jObj.containsKey("sparqlConn")){
			throw new Exception("Cannot create SparqlConnection object because the JSON is wrapped in \"sparqlConn\"");
		}
		
		this.name = (String) jObj.get("name");
		this.domain = (String) jObj.get("domain");
		this.domain = (this.domain == null) ? "" : this.domain;
		
		this.modelInterfaces = new ArrayList<SparqlEndpointInterface>();
		
		this.dataInterfaces = new ArrayList<SparqlEndpointInterface>();
		
		if (jObj.containsKey("dsURL")) {
			
			// super-duper-backwards compatible read
		
			// If any field doesn't exist, presume it exists in the other connection
			this.addModelInterface(	(String)(jObj.get("type")), 
								    (String)(jObj.containsKey("onURL") ? jObj.get("onURL") : jObj.get("dsURL")),
								    (String)(jObj.containsKey("onDataset") ? jObj.get("onDataset") : jObj.get("dsDataset"))
							        );
			this.addDataInterface(	(String)(jObj.get("type")), 
									(String)(jObj.containsKey("dsURL") ? jObj.get("dsURL") : jObj.get("onURL")),
									(String)(jObj.containsKey("dsDataset") ? jObj.get("dsDataset")  : jObj.get("onDataset"))
									);
		} else {
			// normal read
			
			// read model interfaces
	    	for (int i=0; i < ((JSONArray)(jObj.get("model"))).size(); i++) {
	    		JSONObject m = (JSONObject)((JSONArray)jObj.get("model")).get(i);
	    		this.addModelInterface(SparqlEndpointInterface.getInstance(m));
	    	}
	    	// read data interfaces
	    	for (int i=0; i < ((JSONArray)(jObj.get("data"))).size(); i++) {
	    		JSONObject d = (JSONObject)((JSONArray)jObj.get("data")).get(i);
	    		this.addDataInterface(SparqlEndpointInterface.getInstance(d));
	    	}
		}
		
		this.enableOwlImports = (Boolean) jObj.getOrDefault(KEY_ENABLE_OWL_IMPORTS, false);
		
		// no deprecated field-handling
	}

	/**
	 * Override server in all data and model interfaces
	 * @param serverAndPort
	 * @throws Exception
	 */
	public void overrideSparqlServer(String serverAndPort) throws Exception {
		for (SparqlEndpointInterface sei : this.dataInterfaces) {
			sei.setServerAndPort(serverAndPort);
		}
		for (SparqlEndpointInterface sei : this.modelInterfaces) {
			sei.setServerAndPort(serverAndPort);
		}
	}
	
	/**
	 * Create deep copy
	 * @param other
	 * @return
	 * @throws Exception
	 */
	public static SparqlConnection deepCopy(SparqlConnection other) throws Exception {
		return new SparqlConnection(other.toJson().toString());
	}
	
	/**
	 * Instantiate from json text
	 * @param jsonText
	 * @throws Exception
	 */
	public void fromString(String jsonText) throws Exception {
		JSONParser parser = new JSONParser();
		JSONObject jObj = (JSONObject) parser.parse(jsonText);
		fromJson(jObj);

	}
	
	/**
	 * To json string
	 */
	public String toString() {
		return toJson().toString();
	}
	
	/**
	 * Compare
	 * @param other
	 * @param ignoreName
	 * @return
	 */
	public boolean equals(SparqlConnection other, boolean ignoreName) {
		String thisStr = this.toString();
		String otherStr = other.toString();
		if (ignoreName) {
			thisStr = thisStr.replaceFirst(this.name, "NAME");
			otherStr = otherStr.replaceFirst(other.name, "NAME");
		}
		return (thisStr.equals(otherStr));
	}
	
	public void setName (String name) {
		this.name = name;
	}
	
	public void setDomain (String domain) {
		this.domain = (domain == null) ? "" : domain;
	}
	
	public void addModelInterface(SparqlEndpointInterface sei) {
		this.modelInterfaces.add(sei);
	}
	public void addModelInterface(String sType, String url, String graph) throws Exception {
		this.modelInterfaces.add(SparqlEndpointInterface.getInstance(sType, url, graph));
	}
	
	public void addDataInterface(String sType, String url, String graph) throws Exception {
		this.dataInterfaces.add(SparqlEndpointInterface.getInstance(sType, url, graph));
	}
	public void addModelInterface(String sType, String url, String graph, String user, String passwd) throws Exception {
		this.modelInterfaces.add(SparqlEndpointInterface.getInstance(sType, url, graph, user, passwd));
	}
	
	public void addDataInterface(SparqlEndpointInterface sei) {
		this.dataInterfaces.add(sei);
	}
	public void addDataInterface(String sType, String url, String graph, String user, String passwd) throws Exception {
		this.dataInterfaces.add(SparqlEndpointInterface.getInstance(sType, url, graph, user, passwd));
	}
	public int getModelInterfaceCount() {
		return this.modelInterfaces.size();
	}
	
	public SparqlEndpointInterface getModelInterface(int i) {
		return this.modelInterfaces.get(i);
	}
	
	public ArrayList<SparqlEndpointInterface> getModelInterfaces() {
		return this.modelInterfaces;
	}
	
	public void clearInterfaces() {
		this.clearDataInterfaces();
		this.clearModelInterfaces();
	}
	
	public void clearDataInterfaces() {
		this.dataInterfaces = new ArrayList<SparqlEndpointInterface>();
	}
	public void clearModelInterfaces() {
		this.modelInterfaces = new ArrayList<SparqlEndpointInterface>();
	}
	
	public String getDomain() {
		return this.domain;
	}
	
	public String getName() {
		return this.name;
	}
	
	public int getDataInterfaceCount() {
		return this.dataInterfaces.size();
	}
	
	public SparqlEndpointInterface getDataInterface(int i) {
			return this.dataInterfaces.get(i);
	}
	
	public ArrayList<SparqlEndpointInterface> getDataInterfaces() {
		return this.dataInterfaces;
	}
	
	public SparqlEndpointInterface getDefaultQueryInterface() throws Exception {
		if (this.dataInterfaces.size() > 0) {
			return this.dataInterfaces.get(0);
		} else if (this.modelInterfaces.size() > 0) {
			return this.modelInterfaces.get(0);
		} else {
			throw new Exception("This SparqlConnection has no endpoints.");
		}
	}
	
	public SparqlEndpointInterface getInsertInterface() throws Exception {
		return this.dataInterfaces.get(0);
	}
	
	public SparqlEndpointInterface getDeleteInterface() throws Exception {
		return this.dataInterfaces.get(0);
	}
	
	public boolean isOwlImportsEnabled() {
		return enableOwlImports;
	}

	/**
	 * Default is false for backwards-compatibility reasons
	 * @param enableOwlImports
	 */
	public void setOwlImportsEnabled(boolean enableOwlImports) {
		this.enableOwlImports = enableOwlImports;
	}

	/**
	 * Do all endpoints point to the same server
	 * @return
	 */
	public boolean isSingleServerURL() {
		String url = "";
		for (int i=0; i < this.modelInterfaces.size(); i++) {
			String e =  this.modelInterfaces.get(i).getServerAndPort();
			if (url.equals("")) {
				url = e;
			} else if (! e.equals(url)) {
				return false;
			}
		}
		
		// add data interfaces
		for (int i=0; i < this.dataInterfaces.size(); i++) {
			String e =  this.dataInterfaces.get(i).getServerAndPort();
			if (url.equals("")) {
				url = e;
			} else if (! e.equals(url)) {
				return false;
			}
		}
		
		// if there are no serverURLs then false
		if (url.equals("")) {
			return false;
		} else {
			return true;
		}
	}
	
	/**
	 * Throw exception if connection contains multiple ServerURLs
	 * @return
	 */
	public void confirmSingleServerURL() throws Exception {
		String url = "";
		
		// check data interfaces
		for (int i=0; i < this.dataInterfaces.size(); i++) {
			String e =  this.dataInterfaces.get(i).getServerAndPort();
			if (url.equals("")) {
				url = e;
			} else if (! e.equals(url)) {
				throw new Exception("Connection has two different SPARQL endpoints: " + url + " and " + e);
			}
		}
		// check model interfaces
		for (int i=0; i < this.modelInterfaces.size(); i++) {
			String e =  this.modelInterfaces.get(i).getServerAndPort();
			if (url.equals("")) {
				url = e;
			} else if (! e.equals(url)) {
				throw new Exception("Connection has two different SPARQL endpoints: " + url + " and " + e);
			}
		}
		
		// if there are no serverURLs then false
		if (url.equals("")) {
			throw new Exception("Connection has no SPARQL endpoints: " + this.toString());
		} 
	}
	
	/**
	 * get list of graphs for a given serverURL
	 * @param serverURL
	 * @return
	 */
	public ArrayList<String> getAllGraphsForServer(String serverURL) {
		ArrayList<String> ret = new ArrayList<String>();
		
		ret.addAll(this.getDataDatasetsForServer(serverURL));
		
		// add any models that aren't duplicates of data
		ArrayList<String> modelDatasets = this.getModelDatasetsForServer(serverURL);
		for (String ds : modelDatasets) {
			if (! ret.contains(ds)) {
				ret.add(ds);
			}
		}
		
		return ret;
	}
	
	/**
	 *  get list of DATA datasets for a given serverURL
	 * @param serverURL
	 * @return
	 */
	public ArrayList<String> getDataDatasetsForServer(String serverURL) {
		ArrayList<String> ret = new ArrayList<String>();
		
		for (int i=0; i < this.dataInterfaces.size(); i++) {
			SparqlEndpointInterface e =  this.dataInterfaces.get(i);
			if (e.getServerAndPort().equals(serverURL)  &&  ret.indexOf(e.getGraph()) == -1) {
				ret.add(e.getGraph());
			}
		}
		
		return ret;
	}
	
	/**
	 *  get list of MODEL datasets for a given serverURL
	 * @param serverURL
	 * @return
	 */
	public ArrayList<String> getModelDatasetsForServer(String serverURL) {
		ArrayList<String> ret = new ArrayList<String>();
		
		for (int i=0; i < this.modelInterfaces.size(); i++) {
			SparqlEndpointInterface e =  this.modelInterfaces.get(i);
			if (e.getServerAndPort().equals(serverURL) &&  ret.indexOf(e.getGraph()) == -1) {
				ret.add(e.getGraph());
			}
		}
		
		return ret;
	}
	
	/**
	 * Generate a string that uniquely identifies the model connection(s)
	 * @return
	 */
	public String getUniqueModelKey() {
		
		String modelKeys[] = new String[this.getModelInterfaceCount()];
		
		for (int i=0; i < this.getModelInterfaceCount(); i++) {
			modelKeys[i] = this.buildSeiKey(this.getModelInterface(i));
		}
		Arrays.sort(modelKeys);
		
		StringBuilder ret = new StringBuilder();
		ret.append(this.domain + ";");
		for (int i=0; i < modelKeys.length; i++) {
			ret.append(modelKeys[i] + ";");
		}
		ret.append(this.enableOwlImports ? "owlImports;" : "noImports;");
		return ret.toString();
	}
	
	private String buildSeiKey(SparqlEndpointInterface sei) {
		return sei.getServerAndPort() + "-" + sei.getGraph();
	}
	
	/**
	 * Are any of the graphs in this connection also in the Connection that generated sparqlConnKey
	 * @param sparqlConnKey
	 * @return boolean
	 */
	public boolean overlapsSparqlConnKey(String sparqlConnKey) {
		ArrayList<SparqlEndpointInterface> seiList = new ArrayList<SparqlEndpointInterface>();
		seiList.addAll(this.getModelInterfaces());
		seiList.addAll(this.getDataInterfaces());
		
		for (SparqlEndpointInterface sei : seiList) {
			if (sparqlConnKey.contains(this.buildSeiKey(sei) + ";"))
				return true;
		}
		return false;
	}
	
	/**
	 * Generate a string that uniquely identifies the model and data connection(s)
	 * @return
	 */
	public String getUniqueKey() {
		
		String seiKeys[] = new String[this.getModelInterfaceCount() + this.getDataInterfaceCount()];
		int index = 0;
		
		// build a list of keys: one per graph
		for (int i=0; i < this.getModelInterfaceCount(); i++) {
			seiKeys[index++] = this.buildSeiKey(this.getModelInterface(i));
		}
		for (int i=0; i < this.getDataInterfaceCount(); i++) {
			seiKeys[index++] = this.buildSeiKey(this.getDataInterface(i));
		}
		Arrays.sort(seiKeys);
		
		// append keys if they are not duplicates.  everything is followed by ';'
		StringBuilder ret = new StringBuilder();
		ret.append(this.domain + ";");
		for (int i=0; i < seiKeys.length; i++) {
			if (i == 0 || !seiKeys[i].equals(seiKeys[i-1])) {
				ret.append(seiKeys[i] + ";");
			}
		}
		ret.append(this.enableOwlImports ? "owlImports;" : "noImports;");
		return ret.toString();
	}
}
